### PR TITLE
Define enum for known attributes

### PR DESCRIPTION
**Pull request by rachitnigam**
This is in preparation of the upcoming deprecation of the `static` attribute. By making the attributes be defined in a centralized location, we can quickly deprecate things. Another benefit is now we can find all of the locations where a particular attribute is used.

One thing that I can use input on @EclecticGriffin & @sampsyo: The internal attributes are currently defined in the enum but it would be better if we could gensym them so that they are truly internal. Additionally, are there space optimizations we can perform in the future with this representation?

### CODE REVIEW

**Comment by EclecticGriffin** [comment]
Just to clarify, what do you mean by gensym in this context?

**Comment by rachitnigam** [comment]
Using it as a short hand to mean "create a unique, private, inaccessible name", the idea being that if you're not the pass that generated the symbol, you don't get to use it.

**Comment by EclecticGriffin** [comment]
Gotcha! Thanks for clarifying, I'll review the PR with that in mind later today

**Comment by EclecticGriffin** [comment]
In terms of the desire for a gensym approach:

So the challenge we run into is that enum variants don't exist at the type level, so we can't easily make them opaque to only some parts of the compiler.

One option would be to detach non-universal annotations from the IR and have them exist only for a given pass in an auxiliary structure, hashed by something (possibly raw pointers (memory locations)). This would let each pass define any annotations it wants to use and have them automatically cleaned up when it exits and thus inaccessible remain inaccessible for other passes. Doing this would amount to threading an association map with a generic argument through the visitor pattern. It also avoids placing things on the IR so is a bit nicer on space; however not being directly tied to the IR means that you have access it separately which could be annoying.

An alternative might be to create a marker trait for pass-specific attribute enums and have a variant of the general attribute enum be a dyn box of that. You could then downcast through the `Any` trait inside the pass and as long as the enum used by the pass remains private, the attributes should be inaccessible elsewhere. That said, doing so requires pass-writers to use `Any` in a bunch of places and that's pretty cumbersome.

The secret third option is to leave it not actually internal, though that's hardly desirable.

File: calyx-frontend/src/attribute.rs [hunk 1]
```diff
@@ -1,146 +1,141 @@
...
-    fn index(&self, key: S) -> &u64 {
-        let idx = key.into();
-        self.get(idx)
-            .unwrap_or_else(|| panic!("No key `{}` in attribute map", idx))
-    }
-}
+    fn try_from(s: String) -> CalyxResult<Self> {
+        if DEPRECATED_ATTRIBUTES.contains(&s.as_str()) {
+            return Err(Error::malformed_structure(format!(
+                "Attribute {s} is deprecated"
```
Comment (EclecticGriffin) [inline calyx-frontend/src/attribute.rs:113]: We should probably have a message attached to the deprecated attribute message
which explains what should be done instead. For example the static deprecation
might point the user toward the static group construct

File: calyx-frontend/src/attribute.rs [hunk 2]
```diff
@@ -1,146 +1,141 @@
...
+    ST_ID,
+    LOOP,
+    START,
+    END,
+    /// This assignment is dead
+    DEAD,
+
+    /// Unknown attribute. Should not appear in the Calyx codebase.
+    /// Useful for other frontends using Calyx
+    Unknown(Id),
```
[the remark this answers is not in the document]
Comment (rachitnigam) [reply calyx-frontend/src/attribute.rs:60]: Yup, that's what I wanted too but the problem is that I also want `ir::Attributes` struct's methods to arguments of the form `ir::Attribute` which is not feasible if there is no `Unknown` variant. We could make another enum for bool variants, num variants, and unknown variants (as highlighted in #1430) but the fundamental problem of having `ir::Attribute` be large will not be changed.

File: calyx-frontend/src/attribute.rs [hunk 3]
```diff
@@ -1,146 +1,198 @@
...
+            Ok(Attribute::Bool(b))
+        } else if let Ok(n) = NumAttr::from_str(s) {
+            Ok(Attribute::Num(n))
+        } else {
+            // Reject attributes that all caps since those are reserved for internal attributes
+            if s.to_uppercase() == s {
+                return Err(Error::misc(format!("Invalid attribute: {}. All caps attributes are reserved for internal use.", s)));
+            }
+            Ok(Attribute::Unknown(s.into()))
+        }
```
Comment (sampsyo) [inline calyx-frontend/src/attribute.rs:156]: Just marking this as kind of the "essence" of the change: we are recognizing a list of attributes by name and turning them into special enum values, with a fallback variant for all others.

**Review by sampsyo — COMMENTED** [review]
Really nice work on this; it makes a lot of sense to special-case a blessed list of attributes. Seems like there is essentially no downside to this (it prevents errors and admits a more efficient representation).

For the internal attributes: I don't mind the current policy, which is that all-caps names are reserved and can't appear in source code! That actually seems to avoid the weirdness of proper gensym (like, would we append an integer to avoid conflicts with other names that are used?). So I think I'm pretty comfortable proceeding as-is!

**Pull request merged**